examples/dialogs/dialogs/app.py

```diff
@@ -90,12 +90,53 @@ def action_save_file_dialog(self, widget):
         except ValueError:
             self.label.text = "Save file dialog was canceled"
 
+    def action_open_secondary_window(self, widget):
+        # Problems:
+        # - Setting secondary window's `app` manually
+        # - Secondary window's `icon` is not set
+        def close_handler(win):
+            self.window_count -= 1
+            self.set_window_label_text()
+        window = toga.Window(title="New Window {}".format(self.window_count))
+        self.window_count += 1
+        self.set_window_label_text()
+        secondary_label = toga.Label(text="You are in secondary window!")
+        window.content = toga.Box(
+            children=[
+                secondary_label
+            ],
+            style=Pack(
+                flex=1,
+                direction=COLUMN,
+                padding=10
+            )
+        )
+        window.app = self
+        window.on_close = close_handler
+        window.show()
+
+    def exit_handler(self, app, cancel_exit):
```

Comment:
This is a good idea, but we need to differentiate two things here: an event handler that is triggered when an app *is* exiting, and an event handler that is triggered to determine if an app *can* exit. Some platforms do this with two different handlers (macOS uses willClose and didClose); others use a boolean return value from the handler (true means the app can exit). 

Passing in a callable definitely works in the windows case, but I'm not sure it's the best API overall.

Reply:
> I originally thought that we may not need an public facing `on_exit`, but thinking about it some more, I think it's worth having a separate `on_exit` handler - if only because Document based apps (like Podium) don't have a main window, but obviously they still exit.

When I tried to implement Document-based app (Podium) on Windows, I used a main Form that has IsMdiContainer (is Multiple Document Interface Container) set to true for the container window, and set it as the main window; and secondary windows (Forms) that have MdiParent property set to the container Form. What's the best way of handling this difference between Mac and Windows re: main window presence/ absence?

Reply:
So, I'm not 100% sure what the "right" approach is on Windows. 

On macOS, the App runs exactly once; it always has a menu at the application level; and if you haven't got a document open, a file dialog opens prompting you to open a document. Each Document that is open has a primary window.

On Linux, each "document" is a separate instance of the running app. Each document has a main window with a menu.

I haven't used Windows for a while, so I'm not 100% sure what users would expect - whether it's closer to the Apple model (but with a menu on each document window), or Linux, somewhere between the two, or something completely different. I remember MDI apps from the days of Windows 3.1 - but that was a "main" window that contained *internal* sub-windows - but that style of app hasn't been common for a long time, AFAIK. If Winforms concept of MDI is different, it might be the right approach; but it would be helpful to see what that looks like to compare with other Windows apps.

The best suggestion I've got is to step away from Toga for a moment, and try and build an app directly with Python.NET that has the right behavior. The app doesn't need to actually *do* anything - putting up a window with a label that says "I am document N". That way we can make sure we have the basic ideas right before we worry about integrating with Toga.

Reply:
I got interested, and dug into this a little more. If Notepad is any indication, the Linux approach might be the right one. If you open to two documents, there are two instances of notepad.exe. So, that should make things a lot easier - we can keep the same "main" window infrastructure, and wrap it with some file handling. The Linux implementation should also be a fairly good point of reference.

Reply:
I think you are right that Windows is closer to Linux here. One point I would add is usually, when you have a document open in an app, and want to open a new one, Windows doesn't start a new app instance, it just replaces the document in the main window of the previous app. Maybe, @brettcannon could answer with more certainty?..

Reply:
I am most definitely not a Windows expert so I can't really answer what the UX should be here.

Reply:
Oh, sorry about this misunderstanding on my part.